pkg/ccl/sqlproxyccl/forwarder.go

```diff
@@ -130,37 +276,428 @@ func (f *forwarder) Close() {
+func (f *forwarder) prepareTransfer() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.mu.state != stateReady {
+		return errors.New("transfer is already in-progress")
+	}
+	f.mu.transferKey = uuid.MakeV4().String()
+	f.mu.transferCloserCh = make(chan struct{})
+	f.mu.state = stateTransferRequested
+	f.mu.transferCtx = nil
+	f.mu.transferConnRecoverable = false
+	return nil
+}
+
+// finishTransfer moves the forwarder back to the ready state, and closes the
+// transferCloser channel (which unblocks the client-to-server processor). This
+// returns an error if it is called during the steady state.
+//
+// NOTE: This should only be called if the connection is safe to continue
+// because this unblocks the client-to-server processor, which may result in
+// more packets being sent to the server. If the connection is unsafe to
+// proceed, we should just call Close().
+func (f *forwarder) finishTransfer() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.mu.state == stateReady {
+		return errors.New("no transfer in-progress")
+	}
+	f.mu.transferKey = ""
+	// This nil case should not happen, but we'll check to avoid closing nil
+	// channels, which causes panics.
+	if f.mu.transferCloserCh != nil {
+		close(f.mu.transferCloserCh)
+	}
+	f.mu.transferCloserCh = nil
+	f.mu.state = stateReady
+	f.mu.transferCtx = nil
+	f.mu.transferConnRecoverable = false
+	return nil
+}
+
+// processTransfer attempts to perform the connection migration, and blocks until
+// the connection has been migrated, or an error has occurred. If the connection
+// has been migrated successfully, retErr == nil.
+//
+// If retErr != nil, the forwarder has to be closed by the caller to prevent any
+// data corruption, with one exception: the caller may choose to abort the
+// transfer process and continue with the fowarding if and only if the error
+// has been marked with errConnRecoverableSentinel, which can be verified with
+// isConnRecoverableError().
+//
+// NOTE: f.mu.transferCtx has to be set before calling this. We use transferCtx
+// instead of ctx here to ensure that we can recover if we fail to connect, or
+// deserialize the session. Binding to ctx means the only way to abort is to
+// close the forwarder.
+func (f *forwarder) processTransfer() (retErr error) {
+	f.mu.Lock()
+	transferKey, transferCtx := f.mu.transferKey, f.mu.transferCtx
+	f.mu.Unlock()
+
+	if transferCtx == nil {
+		return errors.Wrap(errTransferProtocol, "transferCtx is nil")
+	}
+
+	state, revivalToken, err := waitForShowTransferState(transferCtx,
+		f.serverInterceptor, f.clientConn, transferKey)
+	if err != nil {
+		// Some errors may be recoverable, but those are handled in
+		// awaitTransferStateResponse, and marked accordingly.
+		return err
+	}
+
+	f.mu.Lock()
+	f.mu.transferConnRecoverable = true
+	f.mu.Unlock()
+
+	// Connect to a new SQL pod.
+	//
+	// TODO(jaylim-crl): There is a possibility where the same pod will get
```

Comment:
Consider adding the target pod as an argument to RequestTransfer. That way a central component gets to choose where the connections go.